README.md
# swipe_stack

Swipe stack flutter.

- [swipe_stack](#swipestack)
  - [Screenshot](#Screenshot)
  - [Installation](#Installation)
  - [Usage](#Usage)
  - [Manuel Usage](#Manuel-Usage)

## Screenshot

<img src="https://raw.githubusercontent.com/gmetekorkmaz/images/master/swipe-stack/swipe.jpg" width="300">
<img src="https://raw.githubusercontent.com/gmetekorkmaz/images/master/swipe-stack/top.jpg" width="300">
<img src="https://raw.githubusercontent.com/gmetekorkmaz/images/master/swipe-stack/left.jpg" width="300">
<img src="https://raw.githubusercontent.com/gmetekorkmaz/images/master/swipe-stack/right.jpg" width="300">
<img src="https://raw.githubusercontent.com/gmetekorkmaz/images/master/swipe-stack/bottom.jpg" width="300">

[Youtube Video](https://www.youtube.com/watch?v=vWttY4jWa7k)

## Installation

Add 

```bash
swipe_stack : ^latest_version
```
to your pubspec.yaml.

## Usage

| Parameter | Default | Description |
| --------- | ------- | ----------- |
| maxAngle  | 35 | Maximum rotation value. Must be int between 0 and 360 |
| threshold | 30  | Manual swipe is canceled when the card is dragged less than threshold. Must be int between 1 and 100 |
| stackFrom | StackFrom.None | Stack type. Must be one of StackFrom.Left, StackFrom.Right, StackFrom.Top, StackFrom.Bottom, StackFrom.None |
| visibleCount | 2 | Number of objects to display. Must int and at least 2 |
| translationInterval | 0 | Distance between objects. Must be positive integer |
| scaleInterval | 0 | Decreasing scale value. Must be positive double |
| animationDuration | Duration(milliseconds: 200) | Animation duration |
| historyCount | 1 | Number of objects to keep in the history. Must be positive integer |
| padding | EdgeInsets.symmetric(vertical: 20, horizontal: 25) | Main container padding |

```bash
SwipeStack(
    children: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map((int index) {
        return SwiperItem(
            builder: (SwiperPosition position, double progress) {
                return Material(
                    elevation: 4,
                    borderRadius: BorderRadius.all(Radius.circular(6)),
                    child: Container(
                        decoration: BoxDecoration(
                            color: Colors.white,
                            borderRadius: BorderRadius.all(Radius.circular(6)),
                        ),
                        child: Column(
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: <Widget>[
                                Text("Item $index", style: TextStyle(color: Colors.black, fontSize: 20)),
                                Text("Progress $progress", style: TextStyle(color: Colors.blue, fontSize: 12)),
                            ],
                        )
                    )
                );
            }
        );
    }).toList(),
    visibleCount: 3,
    stackFrom: StackFrom.Top,
    translationInterval: 6,
    scaleInterval: 0.03,
    onEnd: () => debugPrint("onEnd"),
    onSwipe: (int index, SwiperPosition position) => debugPrint("onSwipe $index $position"),
    onRewind: (int index, SwiperPosition position) => debugPrint("onRewind $index $position"),
)
```

## Manuel Usage

```bash
final GlobalKey<SwipeStackState> _swipeKey = GlobalKey<SwipeStackState>();
```

```bash
SwipeStack(
    key: _swipeKey,
    ...,
)
```

```bash
_swipeKey.currentState.swipeLeft();
_swipeKey.currentState.swipeRight();
_swipeKey.currentState.rewind();
```

CHANGELOG.md
## [1.0.0] - TODO: Add release date.

lib/swipe_stack.dart
import 'dart:math' as Math;
import 'package:flutter/widgets.dart';

enum SwiperPosition { None, Left, Right }
enum StackFrom { None, Top, Left, Right, Bottom }

class SwiperItem {

  Widget Function(SwiperPosition, double progress) builder;

  SwiperItem({
    @required
    this.builder
  });
  
}

class SwipeStack extends StatefulWidget {

  final List<SwiperItem> children;
  final int maxAngle;
  final int threshold;
  final StackFrom stackFrom;
  final int visibleCount;
  final int translationInterval;
  final double scaleInterval;
  final Duration animationDuration;
  final int historyCount;
  final void Function(int, SwiperPosition) onSwipe;
  final void Function(int, SwiperPosition) onRewind;
  final void Function() onEnd;
  final EdgeInsetsGeometry padding;

  SwipeStack({
    Key key,
    @required
    this.children,
    this.maxAngle = 35,
    this.threshold = 30,
    this.stackFrom = StackFrom.None,
    this.visibleCount = 2,
    this.translationInterval = 0,
    this.scaleInterval = 0,
    this.animationDuration = const Duration(milliseconds: 200),
    this.historyCount = 1,
    this.onEnd,
    this.onSwipe,
    this.onRewind,
    this.padding = const EdgeInsets.symmetric(vertical: 20, horizontal: 25)
  }) : 
    assert(maxAngle >= 0 && maxAngle <= 360),
    assert(threshold >= 1 && threshold <= 100),
    assert(visibleCount >= 2),
    assert(translationInterval >= 0),
    assert(scaleInterval >= 0),
    assert(historyCount >= 0),
    super(key: key);

  SwipeStackState createState() => SwipeStackState();
}

class SwipeStackState extends State<SwipeStack> with SingleTickerProviderStateMixin {

  AnimationController _animationController;
  Animation<double> _animationX;
  Animation<double> _animationY;
  Animation<double> _animationAngle;
  
  double _left = 0;
  double _top = 0;
  double _angle = 0;
  double _maxAngle = 0;
  double _progress = 0;
  double _centerSlow = 1;
  SwiperPosition _currentItemPosition = SwiperPosition.None;
  final List<Map<String, dynamic>> _history = [];

  final Map<StackFrom, Alignment> _alignment = {
    StackFrom.Left: Alignment.centerLeft,
    StackFrom.Top: Alignment.topCenter,
    StackFrom.Right: Alignment.centerRight,
    StackFrom.Bottom: Alignment.bottomCenter,
    StackFrom.None: Alignment.center
  };

  bool _isTop = false;
  bool _isLeft = false;

  int _animationType = 0;
  // 0 None, 1 move, 2 manuel, 3 rewind

  BoxConstraints _baseContainerConstraints;

  int get currentIndex => widget.children.length - 1;
  
  @override
  void initState() {
    if (widget.maxAngle > 0)
      _maxAngle = widget.maxAngle * (Math.pi / 180);

    _animationController = AnimationController(duration: widget.animationDuration, vsync: this);

    _animationController.addListener(() {
      if (_animationController.status == AnimationStatus.forward) {

        if (_animationX != null)
          _left = _animationX.value;

        if (_animationY != null)
          _top = _animationY.value;

        if (_animationType != 1 && _animationAngle != null)
          _angle = _animationAngle.value;

        _progress = (100 / _baseContainerConstraints.maxWidth) * _left.abs();
        _currentItemPosition = (_left.toInt() == 0) ? SwiperPosition.None : (_left < 0) ? SwiperPosition.Left : SwiperPosition.Right;

        setState(() {});
      }
    });

    _animationController.addStatusListener((AnimationStatus animationStatus) {
      if (animationStatus == AnimationStatus.completed) {
        
        // history
        if (_animationType != 3 && _animationType != 0) {
          if (widget.historyCount > 0) {
            _history.add({
              "item": widget.children[widget.children.length-1],
              "position": _currentItemPosition,
              "left": _left,
              "top": _top,
              "angle": _angle
            });

            if (_history.length > widget.historyCount)
              _history.removeAt(0);
          }
        } else if (_animationType == 3) {
          if (widget.onRewind != null)
            widget.onRewind(widget.children.length-1, _history[_history.length-1]["position"]);
          _history.removeAt(_history.length-1);
        }

        if (_animationType != 0 && _animationType != 3) {
          widget.children.removeAt(widget.children.length-1);

          if (widget.onSwipe != null)
            widget.onSwipe(widget.children.length, _currentItemPosition);

          if (widget.children.length == 0 && widget.onEnd != null)
            widget.onEnd();

        }

        _left = 0;
        _top = 0;
        _angle = 0;
        _progress = 0;
        _currentItemPosition = SwiperPosition.None;
        _animationType = 0;
        setState((){});
        _animationController.reset();
      }
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        _baseContainerConstraints = constraints;

        if (widget.children.length == 0)
          return Container();

        return Container(
          padding: widget.padding,
          child: LayoutBuilder(
            builder: (BuildContext context, BoxConstraints constraints) {
              return Stack(
                overflow: Overflow.visible,
                fit: StackFit.expand,
                children: widget.children.asMap().map((int index, _){
                  return MapEntry(
                    index,
                    _item(constraints, index)
                  );
                }).values.toList(),
              );
            },
          ),
        );
      }
    );
  }

  Widget _item(BoxConstraints constraints, int index) {

    if (index != widget.children.length-1) {

      double scaleReduced = (widget.scaleInterval * (widget.children.length - index));
      scaleReduced -= ((widget.scaleInterval * 2) / 100) * _progress;
      final double scale = 1 - scaleReduced;
      
      double positionReduced = ((widget.translationInterval * (widget.children.length - index - 1))).toDouble();
      positionReduced -= (widget.translationInterval / 100) * _progress;
      final double position = positionReduced * - 1;
      
      return Visibility(
        visible: (widget.children.length - index) <= widget.visibleCount,
        child: Positioned(
          top: widget.stackFrom == StackFrom.Top ? position : null,
          left: widget.stackFrom == StackFrom.Left ? position : null,
          right: widget.stackFrom == StackFrom.Right ? position : null,
          bottom: widget.stackFrom == StackFrom.Bottom ? position : null,
          child: Transform.scale(
            scale: scale,
            alignment: _alignment[widget.stackFrom],
            child: Container(
              constraints: constraints,
              child: widget.children[index].builder(SwiperPosition.None, 0)
            )
          )
        ),
      );
    }

    if (widget.maxAngle > 0 && _animationController.status != AnimationStatus.forward) {
      _angle = ((_maxAngle / 100) * _progress) * _centerSlow;
      _angle = _angle * ((_isTop && _isLeft) ? 1 : (!_isTop && !_isLeft) ? 1 : -1);
    }

    return Positioned(
      left: _left,
      top: _top,
      child: GestureDetector(
        child: Transform.rotate(
          angle: _angle,
          child: Container(
            constraints: constraints,
            child: widget.children[index].builder(_currentItemPosition, _progress)
          ),
        ),
        onPanStart: (DragStartDetails dragStartDetails) {
          RenderBox getBox = context.findRenderObject();
          var local = getBox.globalToLocal(dragStartDetails.globalPosition);

          _isLeft = local.dx < getBox.size.width / 2;
          _isTop = local.dy < getBox.size.height / 2;

          double halfHeight = getBox.size.height / 2;
          _centerSlow = ((halfHeight - local.dy) * (1 / halfHeight)).abs();

        },
        onPanUpdate: (DragUpdateDetails dragUpdateDetails) {
          _left += dragUpdateDetails.delta.dx;
          _top += dragUpdateDetails.delta.dy;

          _progress = (100 / _baseContainerConstraints.maxWidth) * _left.abs();
          _currentItemPosition = (_left.toInt() == 0) ? SwiperPosition.None : (_left < 0) ? SwiperPosition.Left : SwiperPosition.Right;
          setState(() {});
        },
        onPanEnd: _onPandEnd
      ),
    );

  }

  void _onPandEnd(_) {
    setState((){});
    if (_progress < widget.threshold) {
      _goFirstPosition();
    } else {
      _animationType = 1;
      _animationX = Tween<double>(begin: _left, end: _baseContainerConstraints.maxWidth * (_left < 0 ? -1 : 1)).animate(_animationController);
      _animationY = Tween<double>(begin: _top, end: _top + _top).animate(_animationController);
      _animationController.forward();
    }
  }

  void _goFirstPosition() {
    _animationX = Tween<double>(begin: _left, end: 0.0).animate(_animationController);
    _animationY = Tween<double>(begin: _top, end: 0.0).animate(_animationController);
    if (widget.maxAngle > 0)
      _animationAngle = Tween<double>(begin: _angle, end: 0.0).animate(_animationController);
    _animationController.forward();
  }
  
  void swipeLeft() {
    if (widget.children.length > 0 && _animationController.status != AnimationStatus.forward) {
      _animationType = 2;
      _animationX = Tween<double>(begin: 0, end: _baseContainerConstraints.maxWidth * -1).animate(_animationController);
      _animationY = Tween<double>(begin: 0, end: (_baseContainerConstraints.maxHeight / 2) * -1).animate(_animationController);
      if (widget.maxAngle > 0)
        _animationAngle = Tween<double>(begin: 0, end: _maxAngle * 0.7).animate(_animationController);
      _animationController.forward();
    }
  }

  void swipeRight() {
    if (widget.children.length > 0 && _animationController.status != AnimationStatus.forward) {
      _animationType = 2;
      _animationX = Tween<double>(begin: 0, end: _baseContainerConstraints.maxWidth).animate(_animationController);
      _animationY = Tween<double>(begin: 0, end: (_baseContainerConstraints.maxHeight / 2) * -1).animate(_animationController);
      if (widget.maxAngle > 0)
        _animationAngle = Tween<double>(begin: 0, end: (_maxAngle * 0.7) * -1).animate(_animationController);
      _animationController.forward();
    }
  }

  void rewind() {
    if (_history.length > 0 && _animationController.status != AnimationStatus.forward) {
      _animationType = 3;

      final lastHistory = _history[_history.length-1];

      widget.children.add(lastHistory["item"]);
      _animationX = Tween<double>(begin: lastHistory["left"], end: 0).animate(_animationController);
      _animationY = Tween<double>(begin: lastHistory["top"], end: 0).animate(_animationController);
      if (widget.maxAngle > 0)
        _animationAngle = Tween<double>(begin: lastHistory["angle"], end: 0).animate(_animationController);

      _animationController.forward();
    }
  }

  void clearHistory() => _history.clear();

  @override
  void dispose() { 
    _animationController.dispose();
    super.dispose();
  }

}
